airflow/models/dag.py

```diff
@@ -2730,22 +2730,50 @@ def add_logger_if_needed(ti: TaskInstance):
         )
 
         tasks = self.task_dict
+        # Special case when retry is set and DAG is paused
+        task_try_number: dict[str, int] = {}
+
         self.log.debug("starting dagrun")
         # Instead of starting a scheduler, we run the minimal loop possible to check
         # for task readiness and dependency management. This is notably faster
         # than creating a BackfillJob and allows us to surface logs to the user
         while dr.state == State.RUNNING:
             schedulable_tis, _ = dr.update_state(session=session)
-            try:
-                for ti in schedulable_tis:
+
+            for ti in schedulable_tis:
+                try:
                     add_logger_if_needed(ti)
                     ti.task = tasks[ti.task_id]
                     _run_task(ti, session=session)
-            except Exception:
-                self.log.info(
-                    "Task failed. DAG will continue to run until finished and be marked as failed.",
-                    exc_info=True,
-                )
+                except Exception:
+                    if ti.state == TaskInstanceState.UP_FOR_RETRY:
+                        try_number = task_try_number.get(f"{ti.task_id}_{ti.map_index}", 0)
+                        if try_number > ti.max_tries:
+                            ti.set_state(TaskInstanceState.FAILED)
+                        else:
+                            task_try_number[f"{ti.task_id}_{ti.map_index}"] = try_number + 1
+                    self.log.info(
+                        "Task failed. DAG will continue to run until finished and be marked as failed.",
+                        exc_info=True,
+                    )
+            for ti in dr.get_task_instances(session=session):
```

Comment:
Hi @uranusjr thanks for the cleanup. I tested it but with the session param in `get_task_instances` i. e `dr.get_task_instances(session=session)` look like I'm getting the correct TI state below i.e I'm getting deferred state here but in UI it is in the scheduled state, it worked fine when I remove this session param,

Reply:
Hmm, this sounds problematic to me. Removing `session` means the function opens a new database session, which would not contain changes made in the session in the outer scope. If this causes state issue we should probably try to figure out what the cause is.

Reply:
I looked at this, I did not find a reason for this weird behaviour. But In this block, I care only about scheduled state TI so I have changed the function call to fetch only scheduled state TI to move this PR forward, interestingly this is working. which suggests might be something with caching, but not sure.